src/main/io/statusindicator.c

```diff
@@ -20,71 +20,101 @@
+#include "drivers/accgyro.h"
+
+#include "io/serial_cli.h"
+#include "io/rc_controls.h"
+
+#include "sensors/sensors.h"
+#include "sensors/acceleration.h"
+
+#include "flight/failsafe.h"
+#include "flight/imu.h"
+
+#include "mw.h"
+#include "scheduler.h"
 
 #include "statusindicator.h"
 
 static uint32_t warningLedTimer = 0;
 
 typedef enum {
-    WARNING_LED_OFF = 0,
-    WARNING_LED_ON,
-    WARNING_LED_FLASH
-} warningLedState_e;
-
-static warningLedState_e warningLedState = WARNING_LED_OFF;
+    ARM_PREV_NONE       = 0,
+    ARM_PREV_CLI        = 0x00205, //         0b1000000101  2 flashes - CLI active in the configurator
+    ARM_PREV_FAILSAFE   = 0x00815, //       0b100000010101  3 flashes - Failsafe mode
+    ARM_PREV_ANGLE      = 0x02055, //     0b10000001010101  4 flashes - Maximum arming angle exceeded
+    ARM_PREV_CALIB      = 0x08155, //   0b1000000101010101  5 flashes - Calibration active
+    ARM_PREV_OVERLOAD   = 0x20555  // 0b100000010101010101  6 flashes - System overload
+} armingPreventedReason_e;
 
-void warningLedResetTimer(void)
-{
-    uint32_t now = millis();
-    warningLedTimer = now + 500000;
-}
+static uint32_t blinkMask = 0;
 
-void warningLedEnable(void)
+armingPreventedReason_e getArmingPreventionReason(void)
 {
-    warningLedState = WARNING_LED_ON;
+    if (isCalibrating()) {
+        return ARM_PREV_CALIB;
+    }
+    if (rcModeIsActive(BOXFAILSAFE) || failsafePhase() == FAILSAFE_LANDED) {
+        return ARM_PREV_FAILSAFE;
+    }
+    if (!imuIsAircraftArmable(armingConfig()->max_arm_angle)) {
+        return ARM_PREV_ANGLE;
+    }
+    if (cliMode) {
+        return ARM_PREV_CLI;
+    }
+    if (isSystemOverloaded()) {
+        return ARM_PREV_OVERLOAD;
+    }
+    return ARM_PREV_NONE;
 }
 
-void warningLedDisable(void)
-{
-    warningLedState = WARNING_LED_OFF;
-}
 
-void warningLedFlash(void)
+void warningLedRefresh(void)
 {
-    warningLedState = WARNING_LED_FLASH;
+    if (blinkMask <= 1) {  // skip interval for terminator bit, allow continuous on
+        blinkMask = getArmingPreventionReason();
+    }
+    if (blinkMask) {
+        if (blinkMask & 1)
+            LED0_ON;
+        else
+            LED0_OFF;
+        blinkMask >>= 1;
+    }
 }
 
-void warningLedRefresh(void)
+void warningLedBeeper(bool on)
 {
-    switch (warningLedState) {
-        case WARNING_LED_OFF:
-            LED0_OFF;
-            break;
-        case WARNING_LED_ON:
+    if (!blinkMask) {
+        if (on) {
             LED0_ON;
-            break;
-        case WARNING_LED_FLASH:
-            LED0_TOGGLE;
-            break;
+        } else {
+            LED0_OFF;
+        }
     }
-
-    uint32_t now = micros();
-    warningLedTimer = now + 500000;
 }
 
 void warningLedUpdate(void)
 {
     uint32_t now = micros();
-
-    if ((int32_t)(now - warningLedTimer) < 0) {
-        return;
+    if ((int32_t)(now - warningLedTimer) > 0) {
+        warningLedRefresh();
+        warningLedTimer = now + WARNING_LED_BLINK_DELAY;
```

Comment:
One thing, only for reference:
other possibility is to use `warningLedTimer += WARNING_LED_BLINK_DELAY;`. 
That way timing is more accurate (difference between `now` and `warningLedTimer` is not discarded as in current code. It is even possible to use fractional format for time and invocation intervals get rounded correctly.
Big disadvantage is the you must sanitize `warningLedTimer` (called too late/invalid value).

For blinking, timing from last invocation (current code) is usually better (blink lengths are consistent, not absolute time).